drivers/spi/spi_gd32.c

```diff
@@ -195,6 +233,146 @@ static int spi_gd32_frame_exchange(const struct device *dev)
 	return spi_gd32_get_err(cfg);
 }
 
+#ifdef CONFIG_SPI_GD32_DMA
+static void spi_gd32_dma_callback(const struct device *dmadev, void *arg,
+				  uint32_t channel, int status);
+
+static uint32_t spi_gd32_dma_setup(const struct device *dev, uint32_t dir)
+{
+	const struct spi_gd32_config *cfg = dev->config;
+	struct spi_gd32_data *data = dev->data;
+	const uint32_t datareg = (uint32_t)&SPI_DATA(cfg->reg);
+	struct dma_config *dma_cfg = &data->dma[dir].config;
+	struct dma_block_config *block_cfg = &data->dma[dir].block;
+	const struct spi_gd32_dma_config *dma = &cfg->dma[dir];
+	static uint32_t dummy_tx;
+	static uint32_t dummy_rx;
+	int ret;
+
+	memset(dma_cfg, 0, sizeof(struct dma_config));
+	memset(block_cfg, 0, sizeof(struct dma_block_config));
+
+	dma_cfg->source_burst_length = 1;
+	dma_cfg->dest_burst_length = 1;
+	dma_cfg->user_data = (void *)dev;
+	dma_cfg->dma_callback = spi_gd32_dma_callback;
+	dma_cfg->block_count = 1U;
+	dma_cfg->head_block = block_cfg;
+	dma_cfg->dma_slot = cfg->dma[dir].slot;
+	dma_cfg->channel_priority =
+		GD32_DMA_CONFIG_PRIORITY(cfg->dma[dir].config);
+	dma_cfg->channel_direction =
+		dir == TX ? MEMORY_TO_PERIPHERAL : PERIPHERAL_TO_MEMORY;
+
+	if (SPI_WORD_SIZE_GET(data->ctx.config->operation) == 8) {
+		dma_cfg->source_data_size = 1;
+		dma_cfg->dest_data_size = 1;
+	} else {
+		dma_cfg->source_data_size = 2;
+		dma_cfg->dest_data_size = 2;
+	}
+
+	block_cfg->block_size = spi_context_max_continuous_chunk(&data->ctx);
+
+	if (dir == TX) {
+		block_cfg->dest_address = datareg;
+		block_cfg->dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			block_cfg->source_address = (uint32_t)data->ctx.tx_buf;
+			block_cfg->source_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		} else {
+			block_cfg->source_address = (uint32_t)&dummy_tx;
+			block_cfg->source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		}
+	}
+
+	if (dir == RX) {
+		block_cfg->source_address = datareg;
+		block_cfg->source_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+
+		if (spi_context_rx_buf_on(&data->ctx)) {
+			block_cfg->dest_address = (uint32_t)data->ctx.rx_buf;
+			block_cfg->dest_addr_adj = DMA_ADDR_ADJ_INCREMENT;
+		} else {
+			block_cfg->dest_address = (uint32_t)&dummy_rx;
+			block_cfg->dest_addr_adj = DMA_ADDR_ADJ_NO_CHANGE;
+		}
+	}
+
+	ret = dma_config(dma->dev, dma->channel, dma_cfg);
+	if (ret < 0) {
+		LOG_ERR("dma_config %p failed %d\n", dma->dev, ret);
+		return ret;
+	}
+
+	ret = dma_start(dma->dev, dma->channel);
+	if (ret < 0) {
+		LOG_ERR("dma_start %p failed %d\n", dma->dev, ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int spi_gd32_start_dma_transceive(const struct device *dev)
+{
+	const struct spi_gd32_config *cfg = dev->config;
+	struct spi_gd32_data *data = dev->data;
+	const size_t chunk_len = spi_context_max_continuous_chunk(&data->ctx);
+	struct dma_status stat;
+	int ret = 0;
+
+	if (!cfg->dma[TX].dev && spi_context_tx_on(&data->ctx)) {
```

Comment:
Can we move dma device check out of this function?

From below code, spi dma `TX` and `RX` enabled simultaneously.
```
	SPI_CTL1(cfg->reg) |= (SPI_CTL1_DMATEN | SPI_CTL1_DMAREN);
```

Reply:
DMA has been changed to guarantee its existence at initialization, so the check here is no longer necessary.
Removed from here.